spring-webmvc/src/main/java/org/springframework/web/servlet/mvc/method/annotation/ServletRequestMethodArgumentResolver.java

```diff
@@ -84,12 +84,13 @@
 	@Override
 	public boolean supportsParameter(MethodParameter parameter) {
 		Class<?> paramType = parameter.getParameterType();
+		final Annotation[] parameterAnnotations = parameter.getParameterAnnotations();
 		return (WebRequest.class.isAssignableFrom(paramType) ||
 				ServletRequest.class.isAssignableFrom(paramType) ||
 				MultipartRequest.class.isAssignableFrom(paramType) ||
 				HttpSession.class.isAssignableFrom(paramType) ||
 				(pushBuilder != null && pushBuilder.isAssignableFrom(paramType)) ||
-				Principal.class.isAssignableFrom(paramType) ||
+				(Principal.class.isAssignableFrom(paramType) && parameterAnnotations.length == 0) ||
```

Comment:
Merely checking for the absence of an annotation would not be sufficient, since the parameter could be annotated with annotations other than Spring Security's `@AuthenticationPrincipal`.

Reply:
I think the Spring Security issue would need to addressed within Spring Security, potentially making use of `RequestMappingHandlerAdapter.setArgumentResolvers(List<HandlerMethodArgumentResolver>)` (as opposed to `setCustomArgumentResolvers(...)`).

@rstoyanchev and @rwinch, thoughts?

Reply:
I don't think setting the default resolvers is a good way to go.

We have some precedent with arguments of type `Map` where we do check that it isn't annotated in some way, see 5a3ff3521538725695da77d250c97fb49783b10f. We could so something similar here given that `Principal` is also a very high level interface with many possible uses and implementations. What I'm wondering though is why this problem surfaces only now, was there some change in Spring Security or did it never work?

Reply:
Or does it depend on the user Object of the application? 

Under https://github.com/spring-projects/spring-security/issues/4151 it says one option is for UserDetails to not inherit Principal but from what I see it doesn't https://docs.spring.io/spring-security/site/docs/4.2.15.RELEASE/apidocs/org/springframework/security/core/userdetails/UserDetails.html.

Reply:
> We have some precedent with arguments of type Map where we do check that it isn't annotated in some way, see 5a3ff35.

OK. I wasn't aware of that.

Reply:
> Or does it depend on the user Object of the application?

Yes, it depends on the user-specific implementation of `UserDetails`.

> Under [spring-projects/spring-security#4151](https://github.com/spring-projects/spring-security/issues/4151) it says one option is for UserDetails to not inherit Principal but from what I see [UserDetails](https://docs.spring.io/spring-security/site/docs/4.2.15.RELEASE/apidocs/org/springframework/security/core/userdetails/UserDetails.html) doesn't .

Right. The `UserDetails` interface does not extend `Principal` by default, though a developer may of course choose to implement `Principal` if they want (or need) to.

I have implemented in `UserDetails` in several applications of the years, and I have never had it implement `Principal` as well. I also have never seen anyone do that. But... as I mentioned above, it's certainly an option if it makes sense in a given application.

Reply:
Hi guys thanks for your investment in this issue.

to answer @rstoyanchev 
> What I'm wondering though is why this problem surfaces only now, was there some change in Spring Security or did it never work?

My guess is that it's never worked, people use the `@AuthenticationPrincipal` and they get what they asks for: a `Principal`. It's not the expected one, because it's not coming from the right place, but it's a principal and it's good enough for most devs.

But this behavior has always been a problem:
-  [2019 SO post](https://stackoverflow.com/questions/58817175/java-lang-illegalstateexceptioncurrent-user-principal-is-not-of-type-when-i-try): `HttpServletRequest.getUserPrincipal()` contains the `Authentication` (which extends Principal) so it returns it, and it does not extract the Principal from the Authentication. There are multiple solutions for this around SO and most of them say : *Use `org.keycloak.adapters.springsecurity.token.KeycloakAuthenticationToken` in place of `KeycloakPrincipal`* (which is an implementation of `Authentication`) and magically it starts to work, but not for the good reasons.
- [2015 SO post](https://stackoverflow.com/questions/28834154/why-does-authenticationprincipal-return-the-authentication-instead-of-the-princ): Because `HttpServletRequest.getUserPrincipal()` contains the `Authentication` (which extends Principal).

Basically what you can get with the Google search "**Current user principal is not of type**" is related to this issue.

Both of these issues come from the fact that the Principal resolved is from `HttpServletRequest.getUserPrincipal()` instead of `SecurityContext.getAuthentication.getPrincipal()`.

The nasty thing about that is that the `@AuthenticationPrincipal` has not only become not relevant but also missleading, because all class extending Principal can be injected into a `Principal` type (thanks to the spring-framework, the annotation takes no part in this), but as soon as you want to get your implementation of `Principal` you are doomed.